Add NUMA support and optimizations

tlichwala opened the pull request:
Summary:
This pull request adds support for Non-Uniform Memory Access (NUMA) to Apache Traffic Server, enhancing performance on NUMA systems.

Key Changes:
- Added CMake options to enable NUMA support and debugging.
- Introduced new configuration options for NUMA optimizations.
- Enhanced thread and memory management to be NUMA-aware.
- Added RamCacheContainer for cache duplication across NUMA nodes.
- Integrated NUMA debugging utilities.

Performance testing has shown increased throughput, reduced CPU usage, improved latency, decreased UPI bus load

These changes aim to optimize memory access patterns and reduce latency on NUMA systems.

To build the application with NUMA support, use the following CMake command:
cmake -B build -DCMAKE_BUILD_TYPE=Release -DENABLE_MIMALLOC=ON -DENABLE_NUMA=ON -DENABLE_HWLOC=ON -Dmimalloc_DIR="<mimalloc-dir-path>"

zwoop commented:
[approve ci].

ezelkow1 commented:
[approve ci]

zwoop commented:
Nice! The  PR uses the old Debug() statements, which needs to be changed to the new Dbg / Ctl features:

```
../src/iocore/net/Server.cc:247:3: error: use of undeclared identifier 'Debug'
  Debug("numa", "[Server::listen] Attempting to create socket with family: %d, type: %d, protocol: %d", addr.sa.sa_family,
  ^
../src/iocore/net/Server.cc:256:3: error: use of undeclared identifier 'Debug'
  Debug("numa", "[Server::listen] Attempting to set up fd for listen with non_blocking: %d, options: %d", non_blocking, opt);
  ^
```

cmcfarlen commented on src/cripts/Lulu.cc:
```diff
@@ -163,6 +163,12 @@ cripts::Splitter(cripts::string_view input, char delim)
+{
+  return details::splitter<Cript::string_view>(input, delim);
+}
+
```
You can remove this chunk, perhaps a merge conflict issue.

cmcfarlen commented on src/iocore/net/Server.cc:
```diff
@@ -88,49 +104,205 @@ Server::close()
+    int my_thread_id = this_ethread()->id;
+    int my_numa_node = this_ethread()->get_numa_node();
+
+    Debug("numa_sequencer", "[NUMASequencer] Thread %d (NUMA node %d) entered run_sequential.", my_thread_id, my_numa_node);
```
Debug has been removed, please use the Dbg macro instead.

cmcfarlen reviewed (changes requested):
Thanks so much for this PR! It looks really interesting.  Just a few comments to get ci happy.  Mainly migrating Debug->Dbg.  Looking forward to trying this out!

cmcfarlen commented on src/iocore/cache/RamCacheContainer.cc:
```diff
@@ -0,0 +1,322 @@
+}
+
+RamCache *
+RamCacheContainer::get_cache(unsigned int my_node, unsigned int node)
```
```suggestion
RamCacheContainer::get_cache(unsigned int /* my_node ATS_UNUSED */, unsigned int node)
```

cmcfarlen commented on src/iocore/cache/RamCacheContainer.cc:
```diff
@@ -0,0 +1,322 @@
+
+// returns true if consistent
+static bool
+check_pages_consistency(void *data, size_t size, const char *name = "")
```
This function is unused unless `NUMA_CONSISTENCY_CHECK` is defined.  Please add precompile check here.

cmcfarlen commented on src/iocore/cache/RamCacheContainer.cc:
```diff
@@ -0,0 +1,322 @@
+}
+
+static void
+move_pages_to_current_numa_zone(void *data, size_t size)
```
This function is unused

cmcfarlen commented on src/iocore/net/Server.cc:
```diff
@@ -88,49 +104,205 @@ Server::close()
+
+    // If all threads have been added (assuming their number is equal to eventProcessor.net_threads), sort the thread IDs and set
+    // ready_to_run to true
+    if (thread_ids.size() == eventProcessor.net_threads) {
```
I get a signed/unsigned compare error for this line.

traeak commented:
[approve ci autest 1]

traeak commented:
[approve ci autest1]

traeak commented:
[approve ci autest1]

cmcfarlen commented on include/iocore/eventsystem/EThread.h:
```diff
@@ -691,6 +698,17 @@ EThread::Metrics::decay() -> self_type &
+    getcpu(&cpu, &node);
+    this->numa_node = node;
+  }
+  return this->numa_node;
```
This should handle when numa.h is not available

tlichwala replied:
The getcpu() function is included in the numactl package. Adding a conditional directive should solve this issue.

cmcfarlen reviewed (changes requested):
The build is failing for systems without numa.h.  I noted one place, but there could be other similar issues.

tlichwala commented:
Hi all,
I've addressed the feedback and made the necessary changes. However, the automated tests are still showing build errors on some operating systems. Could someone please help review these issues?
Thank you!

cmcfarlen commented:
It looks like its still related to when numa is not available:
```
../src/iocore/net/QUICPacketHandler.cc: In member function 'virtual void QUICPacketHandlerIn::_recv_packet(int, UDPPacket*)':
../src/iocore/net/QUICPacketHandler.cc:270:65: error: no matching function for call to 'EventProcessor::assign_thread(const int&)'
  270 |     eth                           = eventProcessor.assign_thread(ET_NET);
      |                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
In file included from ../src/iocore/net/../eventsystem/P_EventSystem.h:46,
                 from ../src/iocore/net/P_Net.h:92,
                 from ../src/iocore/net/QUICPacketHandler.cc:25:
../src/iocore/net/../eventsystem/P_UnixEventProcessor.h:54:1: note: candidate: 'EThread* EventProcessor::assign_thread(EventType, int)'
   54 | EventProcessor::assign_thread(EventType etype, int numa_node)
      | ^~~~~~~~~~~~~~
../src/iocore/net/../eventsystem/P_UnixEventProcessor.h:54:1: note:   candidate expects 2 arguments, 1 provided
```

tlichwala commented on include/tscore/ink_thread.h:
```diff
@@ -30,7 +30,15 @@
 #include <sched.h>
+#if TS_USE_HWLOC
+#include <hwloc.h>
+#include <hwloc/glibc-sched.h>
```
I am unable to reproduce the compilation error that occurs when including the header file <hwloc/glibc-sched.h> on freebsd. The error message is: "Please make sure to include sched.h before including glibc-sched.h, and define _GNU_SOURCE before any inclusion of sched.h". I added the undesirable definition of the _GNU_SOURCE macro in the code, as shown below, to ensure that it is defined.

cmcfarlen replied:
hmm, I'm not super familiar with freebsd, but I think this should be added to CXXFLAGS as part of the build setup, not in the source code.  Is Freebsd able to find the numa library?  This feature could be disabled for that OS for now.

tlichwala replied:
The _GNU_SOURCE macro is defined in the CMakeLists.txt file as follows:
list(APPEND CMAKE_REQUIRED_DEFINITIONS -D_GNU_SOURCE -DATS_BUILD)
I added the macro in the source code only for testing purposes to ensure that it is indeed defined.
It appears that the error I am encountering is not related to the numa.h library but rather to hwloc.
Thank you for your suggestion. If you have any further insights or recommendations, I would appreciate it.

cmcfarlen commented:
You might need to add more precompiler conditions for `TS_USE_NUMA` to keep the errors down for systems that don't support this.

cmcfarlen commented:
Unfortunately CMakeLists.txt is conflicting.  You can just accept both the new and old line when resolving.

cmcfarlen commented:
@tlichwala Are you planning on continuing to work on this? Feel free to hit me up on slack to discuss these remaining issues.  I think this work is valuable so don't want it to get too stale.  Thanks!

tlichwala commented:
> @tlichwala Are you planning on continuing to work on this? Feel free to hit me up on slack to discuss these remaining issues. I think this work is valuable so don't want it to get too stale. Thanks!

> Hi Chris, the code is ready for merging, but we currently lack the resources to address the FreeBSD test issues.

cmcfarlen commented:
Bump
